code/cdw.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+CODE DESIGN WORKSHOP 2021
+Functions for pipeline steps
+
+@author:  jenni.saaristo@helsinki.fi
+@version: 2021-04-28
+@notes:   Redo after feedback
+"""
+
+from scipy import signal, stats
+from scipy.spatial.distance import pdist, squareform, cdist
+import numpy as np
+import pandas as pd
+import nibabel as nib
+from itertools import compress
+
+
+def preprocess(bold, labels):
+    """
+    Minimal preprocessing of BOLD data
+    
+    Given a 4D BOLD-data detrends and z-scores it
+    run-wise ("chunks" in labels).
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Original BOLD data.
+    labels : DataFrame
+        Labels for MRI frames.
+
+    Returns
+    -------
+    bold_prep : Nifti1Image
+        Preprocessed data.
+    """
+    
+    print('Copying data...')
+    prep = bold.get_fdata()
+    
+    print('Done. Begin preprocessing...')
+    for c in labels.chunks.unique():
+        print('Processing chunk ' + str(c))
+        inds = labels.query('chunks == @c').index
+        this_prep = prep[:,:,:,inds]
+        this_prep = signal.detrend(this_prep)
+        this_prep = stats.zscore(this_prep, axis=-1)
+        
+        # scipy z-score does not watch out for zero-division, so
+        # we have to correct those ourselves. Replacing nans with
+        # zeros makes sense as long as the input is all zeros, too.
+        if prep[:,:,:,inds][np.isnan(this_prep)].sum() == 0:
+            this_prep[np.isnan(this_prep)] = 0
+        else:
+            raise Exception('Unexpected nans! Aborting.')
+        
+        prep[:,:,:,inds] = this_prep
+    
+    bold_prep = nib.Nifti1Image(prep, bold.affine, bold.header)
+        
+    print('Done.')
+    return bold_prep
+
+def drop_by_labels(bold, labels, bad_labels):
+    """
+    Drop unwanted labels and fMRI frames
+    
+    Takes a 4D BOLD image and drops frames that correspond to unwanted
+    labels, and also drops those labels from the labels DataFrame.
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Full 4D BOLD image.
+    labels : DataFrame
+        Full list of frame labels.
+    bad_labels : list of str
+        Unwanted labels.
+
+    Returns
+    -------
+    bold_subset : Nifti1Image
+        Selected BOLD frames.
+    labels_subset : DataFrame
+        Selected frame labels.
+
+    """
+    
+    good_inds = labels.query('labels not in @bad_labels').index
+    labels_subset = labels.loc[good_inds]
+    
+    print('Dropping unwanted frames...')
+    frames_subset = bold.get_fdata()[:,:,:,good_inds]
```

Comment:
A matter of taste, really, but since `...` is kind of an obscure NumPy feature, maybe you didn't know it existed:

```suggestion
    frames_subset = bold.get_fdata()[..., good_inds]
```

Reply:
Ohhhh NICE. Thanks!

Reply:
In Susan's opinion, this makes the code slightly less readable. Maybe it does. 🤷‍♂️